auth.py

```diff
@@ -207,7 +208,6 @@ def signup ():
         # We automatically login the user
         cookie = make_salt ()
         db_set ('tokens', {'id': cookie, 'username': user ['username'], 'ttl': times () + session_length})
-        db_set ('users', {'username': user ['username'], 'last_login': timems ()})
```

Comment:
This is fine as a fix for now, but there is a deeper underlying problem that `db_set` is inherently risky with its `get+write` behavior.

It's better to make a distinction between "put record" and "update record" from the get-go, and not have it switch between those dynamically (and, as proven here, unpredictably).

The advantage of not having `db_set` do a `db_get` first is also that we'd remove unnecessary calls to DynamoDB on every write.

------

> When logging in a just signed up user, if the timing was very slow, we would overwrite most of the fields of a freshly created user with just username and last_login. 

It sounds like the problem was that a user would log in before their signup record had been written. How could this happen though? I would think that if a user `POST`s to the `/signup` page, we would first do a `db_set` creating the user record and AFTERWARDS redirects with a cookie to the page that does the login?

Or was there a different issue?

> This happened twice.

That explains why we were getting `KeyNotFound` `'email'` errors. Did you remove the broken records from the database?

Reply:
Thanks for the analysis @rix0rrr! Do you think we can merge this for now or is it better to implement your more sustainable fix separating put and update?